sirepo/template/webcon.py

```diff
@@ -356,6 +356,205 @@ def get_settings_report(run_dir, data):
+    x_var = 'x'
+    x_range_var = 'x_range'
+    x_range_inds_var = 'x_range_inds'
+    j = 0
+    rpt_name = 'analysisReport'
+    while rpt_name in data.models:
+        rpt = data.models[rpt_name]
+        x_index = _safe_index(col_info, int(rpt.x))
+        x_label = col_info.names[x_index]
+        y_info = []
+        x_range = None
+        if 'trimField' in rpt and int(rpt.trimField) == x_index:
+            x_range = (float(rpt.trimMin), float(rpt.trimMax))
+        # many params like style are hard-coded - should move to schema?
+        i = 1
+        y_var = f"y{i}"
+        while y_var in rpt:
+            if rpt[y_var] == 'none':
+                i = i + 1
+                y_var = f"y{i}"
+                continue
+            y_index = _safe_index(col_info, int(rpt[y_var]))
+            y_label = col_info.names[y_index]
+            y_info.append(PKDict(
+                y_var=y_var,
+                y_index=y_index,
+                y_label=y_label,
+                style='line'
+            ))
+            i = i + 1
+            y_var = f"y{i}"
+        x_points = f"{data_var}[:, {x_index}]" if not x_range else \
+            f"[x for x in {data_var}[:, {x_index}] if x >= {x_range[0]} and \
+                x <= {x_range[1]}]"
+        x_range_inds = f"([i for (i, x) in enumerate({x_var}) if \
+            x >= {x_range_var}[0]][0], \
+         [i for (i, x) in enumerate({x_var}) if x <= {x_range_var}[1]][-1] + 1)" if \
+            x_range else f"(0, len({x_var}))"
+        code = [
+            f"{x_var} = {x_points}",
+            f"{x_range_var} = {x_range}",
+            f"{x_range_inds_var} = {x_range_inds}",
+        ]
+        for cfg in y_info:
+            code.append(f"{cfg.y_var} = \
+            {data_var}[:, {cfg.y_index}][{x_range_inds_var}[0]:{x_range_inds_var}[1]]")
+        nb.add_code_cell(code)
+        if j == 0:
+            # only do this once
+            nb.add_report(PKDict(
+                x_var=x_var,
+                x_label=x_label,
+                y_info=y_info,
+                title='Analysis Plot'
+            ))
+
+        if 'action' in rpt:
+            # reset y_var to 1st data set
+            y_var = 'y1'
+            nb.add_markdown_cell([f"## {rpt.action} Plot"])
+            if rpt.action == 'fft':
+                # only 1 curve (for now?)
+                w_var = 'w'
+                y_norm_var = 'y_norm'
+                nb.add_code_cell([
+                    f"{w_var}, {y_norm_var} = get_fft({x_var}, {y_var})"
+                ])
+                y_info = [PKDict(
+                    y_var=y_norm_var,
+                    y_label='',
+                    style='line'
+                )]
+                nb.add_report(PKDict(
+                    x_var=w_var,
+                    x_label=x_label,
+                    y_info=y_info,
+                    title='FFT'
+                ))
+            if rpt.action == 'fit':
+                x_fit_var = 'x_fit'
```

Comment:
Too many variables. Try to use more functions, e.g. y_max_var could be _var('y_fit', 'max'), or something.

Reply:
6c893f9